### ISSUE TITLE
providerRef is assumed to be non-nil throughout reconcilers

### ISSUE THREAD

**Issue by displague** [crossplaneio/crossplane#558]
It is possible in Crossplane resources, especially between alpha versions, for a spec property to have been renamed or otherwise not present.

This has been observed in particular for the `providerRef` field.

Because `providerRef` is not safe to assume as non-null, any locations throughout the project that references `instance.spec.ProviderRef` should do so in a null-safe way. 

For example:

```
$ git grep -n NamespacedNameOf | grep Provider
pkg/controller/aws/cache/replicationgroup.go:240:       n := meta.NamespacedNameOf(g.Spec.ProviderReference)
pkg/controller/aws/compute/eks.go:135:  err := r.Get(ctx, meta.NamespacedNameOf(instance.Spec.ProviderReference), p)
pkg/controller/aws/rds/rdsinstance.go:139:      err := r.Get(ctx, meta.NamespacedNameOf(instance.Spec.ProviderReference), p)
pkg/controller/aws/s3/s3bucket.go:139:  err := r.Get(ctx, meta.NamespacedNameOf(instance.Spec.ProviderReference), p)
pkg/controller/azure/cache/redis.go:200:        n := meta.NamespacedNameOf(r.Spec.ProviderReference)
pkg/controller/azure/compute/akscluster.go:166: err := r.Get(ctx, meta.NamespacedNameOf(instance.Spec.ProviderReference), p)
pkg/controller/azure/database/sqlserver.go:73:  n := meta.NamespacedNameOf(instance.GetSpec().ProviderReference)
pkg/controller/azure/resourcegroup/resourcegroup.go:154:        n := meta.NamespacedNameOf(r.Spec.ProviderReference)
pkg/controller/azure/storage/account/account.go:136:    n := meta.NamespacedNameOf(b.Spec.ProviderReference)
pkg/controller/gcp/cache/cloudmemorystore_instance.go:180:      n := meta.NamespacedNameOf(i.Spec.ProviderReference)
pkg/controller/gcp/compute/gke.go:159:  err := r.Get(ctx, meta.NamespacedNameOf(instance.Spec.ProviderReference), p)
pkg/controller/gcp/database/cloudsql_instance.go:154:   n := meta.NamespacedNameOf(instance.Spec.ProviderReference)
pkg/controller/gcp/storage/bucket.go:133:       if err := m.Get(ctx, meta.NamespacedNameOf(b.Spec.ProviderReference), p); err != nil {
```

```
2019-07-01T15:07:34.134Z        DEBUG   crossplane.controller.gke.compute.gcp.crossplane.io     reconciling     {"kind": "GKECluster.compute.gcp.crossplane.io/v1alpha1", "request": "complex/kubernetescluster-6db33c43-981b-11e9-b414-080027b33806"}
E0701 15:07:34.134670       1 runtime.go:69] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
/home/marques/.local/share/go/src/github.com/crossplaneio/crossplane/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:76
/home/marques/.local/share/go/src/github.com/crossplaneio/crossplane/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:65
/home/marques/.local/share/go/src/github.com/crossplaneio/crossplane/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:51
/snap/go/3947/src/runtime/panic.go:522
/snap/go/3947/src/runtime/panic.go:82
/snap/go/3947/src/runtime/signal_unix.go:390
/home/marques/.local/share/go/src/github.com/crossplaneio/crossplane/pkg/controller/gcp/compute/gke.go:159
/home/marques/.local/share/go/src/github.com/crossplaneio/crossplane/pkg/controller/gcp/compute/gke.go:159
/home/marques/.local/share/go/src/github.com/crossplaneio/crossplane/pkg/controller/gcp/compute/gke.go:254
/home/marques/.local/share/go/src/github.com/crossplaneio/crossplane/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:215
/home/marques/.local/share/go/src/github.com/crossplaneio/crossplane/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:158
/home/marques/.local/share/go/src/github.com/crossplaneio/crossplane/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133
/home/marques/.local/share/go/src/github.com/crossplaneio/crossplane/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134
/home/marques/.local/share/go/src/github.com/crossplaneio/crossplane/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88
/snap/go/3947/src/runtime/asm_amd64.s:1337
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x1a04e4e]

goroutine 591 [running]:
github.com/crossplaneio/crossplane/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
        /home/marques/.local/share/go/src/github.com/crossplaneio/crossplane/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:58 +0x105
panic(0x1c585e0, 0x3b28fe0)
        /snap/go/3947/src/runtime/panic.go:522 +0x1b5
github.com/crossplaneio/crossplane/pkg/meta.NamespacedNameOf(...)
        /home/marques/.local/share/go/src/github.com/crossplaneio/crossplane/pkg/meta/meta.go:84
github.com/crossplaneio/crossplane/pkg/controller/gcp/compute.(*Reconciler)._connect(0xc000750ba0, 0xc00038ac00, 0x22c9100, 0xc00038ac00, 0x0, 0x0)
        /home/marques/.local/share/go/src/github.com/crossplaneio/crossplane/pkg/controller/gcp/compute/gke.go:159 +0x5e
github.com/crossplaneio/crossplane/pkg/controller/gcp/compute.(*Reconciler).Reconcile(0xc000750ba0, 0xc000539099, 0x7, 0xc00092ab80, 0x36, 0x3b45c00, 0x0, 0x0, 0x0)
        /home/marques/.local/share/go/src/github.com/crossplaneio/crossplane/pkg/controller/gcp/compute/gke.go:254 +0x335
github.com/crossplaneio/crossplane/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc0000b7ae0, 0x0)
        /home/marques/.local/share/go/src/github.com/crossplaneio/crossplane/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:215 +0x1cc
github.com/crossplaneio/crossplane/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1()
        /home/marques/.local/share/go/src/github.com/crossplaneio/crossplane/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:158 +0x36
github.com/crossplaneio/crossplane/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc001442a80)
        /home/marques/.local/share/go/src/github.com/crossplaneio/crossplane/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x54
github.com/crossplaneio/crossplane/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc001442a80, 0x3b9aca00, 0x0, 0x1, 0xc000594de0)
        /home/marques/.local/share/go/src/github.com/crossplaneio/crossplane/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134 +0xf8
github.com/crossplaneio/crossplane/vendor/k8s.io/apimachinery/pkg/util/wait.Until(0xc001442a80, 0x3b9aca00, 0xc000594de0)
        /home/marques/.local/share/go/src/github.com/crossplaneio/crossplane/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88 +0x4d
created by github.com/crossplaneio/crossplane/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start
        /home/marques/.local/share/go/src/github.com/crossplaneio/crossplane/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:157 +0x311
```

**Comment by ichekrygin** [comment]
IIRC `providerRef` is (was, or maybe should be) a required property in the resource `Spec`

**Comment by ichekrygin** [comment]
@displague after looking into the stack trace, I suspect the issue may be is not that `providerRef` is `nil`, but rather that `providerRef` has an empty `Namespace` value? I believe I stumble over this issue in my latest PR https://github.com/crossplaneio/crossplane/pull/556
```
...
providerRef must include namespace since it was converted from LocalObjectReferece to ObjectReference in the previous PR's
```

**Comment by ichekrygin** [comment]
It will be also very helpful to attach resource `CRD`

**Comment by ichekrygin** [comment]
I was able to repro the issue. The "root" cause is (as it appears) missing `json` tag on `ResourceSpec` field.
```go
type CloudsqlInstanceSpec struct {
	corev1alpha1.ResourceSpec
```
vs.
```go
type CloudsqlInstanceSpec struct {
	corev1alpha1.ResourceSpec `json:",inline"`
```
If `json` tag is omitted the entire `ResourceSpec` appears to be treated as "optional" fields, hence by the extension all the `ResourceSpec` fields, include the required ones.

**Comment by displague** [comment]
Good find, @ichekrygin.

In this case, I think `providerRef` is not present because the resource yaml used in the existing Kubernetes object used the `providerReference` field instead of `providerRef`.  This was a temporary name change in the #545 PR.  My concern is that user input changes, or spec changes from Crossplane, should not trigger stack nil pointer errors.

```json
{
  "kind": "GKECluster.compute.gcp.crossplane.io/v1alpha1",
  "instance": {
    "kind": "GKECluster",
    "apiVersion": "compute.gcp.crossplane.io/v1alpha1",
    "metadata": {
      "name": "kubernetescluster-6db33c43-981b-11e9-b414-080027b33806",
      "namespace": "complex",
      "selfLink": "/apis/compute.gcp.crossplane.io/v1alpha1/namespaces/complex/gkeclusters/kubernetescluster-6db33c43-981b-11e9-b414-080027b33806",
      "uid": "fb4f4bde-981e-11e9-b414-080027b33806",
      "resourceVersion": "462285",
      "generation": 40,
      "creationTimestamp": "2019-06-26T14:30:44Z",
      "ownerReferences": [
        {
          "apiVersion": "compute.crossplane.io/v1alpha1",
          "kind": "KubernetesCluster",
          "name": "wordpress-demo-cluster",
          "uid": "6db33c43-981b-11e9-b414-080027b33806",
          "controller": true
        }
      ],
      "finalizers": [
        "finalizer.gke.compute.gcp.crossplane.io"
      ]
    },
    "spec": {
      "writeConnectionSecretToRef": {},
      "providerRef": null,
      "reclaimPolicy": "Delete",
      "machineType": "n1-standard-1",
      "numNodes": 1,
      "zone": "us-central1-b"
    },
    "status": {
      "conditions": [
        {
          "type": "Ready",
          "status": "True",
          "lastTransitionTime": "2019-06-26T14:34:22Z",
          "reason": "Managed resource is available for use"
        },
        {
          "type": "Synced",
          "status": "True",
          "lastTransitionTime": "2019-06-30T13:21:48Z",
          "reason": "Successfully reconciled managed resource"
        }
      ],
      "bindingPhase": "Bound",
      "clusterName": "gke-fb4f4bde-981e-11e9-b414-080027b33806",
      "endpoint": "35.238.111.61",
      "state": "RUNNING"
    }
  },
  "provider": {
    "metadata": {
      "creationTimestamp": null
    },
    "spec": {
      "credentialsSecretRef": {
        "key": ""
      },
      "projectID": ""
    }
  },
  "context": "context.Background"
}
```

**Comment by ichekrygin** [comment]
I am creating a separate PR to address `json` tag issue.

**Comment by ichekrygin** [comment]
Closed via #559
